aggregator/leader_flush_mgr.go

```diff
@@ -0,0 +1,341 @@
+	sync.RWMutex
+
+	nowFn                  clock.NowFn
+	checkEvery             time.Duration
+	jitterEnabled          bool
+	workers                xsync.WorkerPool
+	flushTimesKeyFmt       string
+	flushTimesStore        kv.Store
+	flushTimesPersistEvery time.Duration
+	logger                 xlog.Logger
+	scope                  tally.Scope
+
+	rand                *rand.Rand
+	flushTimes          flushMetadataHeap
+	flushTimesKey       string
+	lastPersistAt       time.Time
+	flushedSincePersist bool
+	flushTask           *leaderFlushTask
+	metrics             leaderFlushManagerMetrics
+}
+
+func newLeaderFlushManager(opts FlushManagerOptions) roleBasedFlushManager {
+	nowFn := opts.ClockOptions().NowFn()
+	rand := rand.New(rand.NewSource(nowFn().UnixNano()))
+	instrumentOpts := opts.InstrumentOptions()
+	scope := instrumentOpts.MetricsScope()
+	mgr := &leaderFlushManager{
+		nowFn:                  nowFn,
+		checkEvery:             opts.CheckEvery(),
+		jitterEnabled:          opts.JitterEnabled(),
+		workers:                opts.WorkerPool(),
+		flushTimesKeyFmt:       opts.FlushTimesKeyFmt(),
+		flushTimesStore:        opts.FlushTimesStore(),
+		flushTimesPersistEvery: opts.FlushTimesPersistEvery(),
+		logger:                 instrumentOpts.Logger(),
+		scope:                  scope,
+		rand:                   rand,
+		lastPersistAt:          nowFn(),
+		metrics:                newLeaderFlushManagerMetrics(scope),
+	}
+	mgr.flushTask = &leaderFlushTask{mgr: mgr}
+	return mgr
+}
+
+func (mgr *leaderFlushManager) Open(shardSetID string) {
+	mgr.Lock()
+	mgr.flushTimesKey = fmt.Sprintf(mgr.flushTimesKeyFmt, shardSetID)
+	mgr.Unlock()
+}
+
+// Init initializes the leader flush manager by enqueuing all
+// the flushers in the buckets.
+func (mgr *leaderFlushManager) Init(buckets []*flushBucket) {
+	mgr.Lock()
+	mgr.flushTimes.Reset()
+	for bucketIdx, bucket := range buckets {
+		mgr.enqueueBucketWithLock(bucketIdx, bucket)
+	}
+	mgr.Unlock()
+}
+
+func (mgr *leaderFlushManager) Prepare(buckets []*flushBucket) (flushTask, time.Duration) {
+	var (
+		shouldFlush = false
+		duration    tally.Timer
+		flushers    []PeriodicFlusher
+		waitFor     = mgr.checkEvery
+	)
+	mgr.Lock()
+	numFlushTimes := mgr.flushTimes.Len()
+	mgr.metrics.queueSize.Update(float64(numFlushTimes))
+	if numFlushTimes > 0 {
+		earliestFlush := mgr.flushTimes.Min()
+		if nowNanos := mgr.nowNanos(); nowNanos >= earliestFlush.timeNanos {
+			// NB(xichen): make a shallow copy of the flushers inside the lock
+			// and use the snapshot for flushing below.
+			shouldFlush = true
+			bucketIdx := earliestFlush.bucketIdx
+			duration = buckets[bucketIdx].duration
+			flushers = buckets[bucketIdx].flushers
+			nextFlushMetadata := flushMetadata{
+				timeNanos: earliestFlush.timeNanos + int64(buckets[bucketIdx].interval),
+				bucketIdx: bucketIdx,
+			}
+			mgr.flushTimes.Pop()
+			mgr.flushTimes.Push(nextFlushMetadata)
+			mgr.flushedSincePersist = true
```

Comment:
Shouldn't `waitFor` be set to 0 in this code path?  Otherwise how do you ever catch up if you're more than one bucket behind?  

For instance if you ever have more than one bucket waiting to be flushed you'll enter this `nowNanos >= earliestFlush.timeNanos` and do a single flush task but return `waitFor` to be the the `checkEvery` value.  This means you'll never clear out your backlog.